dgraph/cmd/graphql/admin.go

```diff
@@ -0,0 +1,272 @@
+	mutExec := resolve.DgoAsMutationExecutor(dg)
+
+	return resolve.NewResolverFactory().
+		WithQueryResolver("health",
+			func(q schema.Query) resolve.QueryResolver { return &healthResolver{} }).
+		WithMutationResolver("addSchema",
+			func(m schema.Mutation) resolve.MutationResolver {
+				addResolver := &addSchemaResolver{
+					gqlServer:            gqlServer,
+					dgraph:               dg,
+					baseMutationRewriter: mutRw,
+					baseMutationExecutor: mutExec,
+					fns:                  fns,
+					withIntrospection:    true}
+
+				return resolve.NewMutationResolver(
+					addResolver,
+					qryExec,
+					addResolver,
+					resolve.StdMutationCompletion(m.Name()))
+			}).
+		WithQueryResolver("querySchema",
+			func(q schema.Query) resolve.QueryResolver {
+				return resolve.NewQueryResolver(
+					qryRw,
+					qryExec,
+					resolve.StdQueryCompletion())
+			}).
+		WithSchemaIntrospection()
+}
+
+func (asr *addSchemaResolver) Rewrite(m schema.Mutation) (*gql.GraphQuery, []*dgoapi.Mutation, error) {
+	sch, err := getSchemaInput(m)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	schHandler, err := schema.NewHandler(sch)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	asr.newGQLSchema, err = schema.FromString(schHandler.GQLSchema())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	asr.newDgraphSchema = schHandler.DGSchema()
+
+	m.SetArgTo(schema.InputArgName, map[string]interface{}{"schema": sch, "date": time.Now()})
+	return asr.baseMutationRewriter.Rewrite(m)
+}
+
+func (asr *addSchemaResolver) FromMutationResult(
+	mutation schema.Mutation,
+	assigned map[string]string,
+	mutated map[string][]string) (*gql.GraphQuery, error) {
+	return asr.baseMutationRewriter.FromMutationResult(mutation, assigned, mutated)
+}
+
+func (asr *addSchemaResolver) Mutate(
+	ctx context.Context,
+	query *gql.GraphQuery,
+	mutations []*dgoapi.Mutation) (map[string]string, map[string][]string, error) {
+
+	assigned, mutated, err := asr.baseMutationExecutor.Mutate(ctx, query, mutations)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if glog.V(3) {
+		glog.Infof("Altering Dgraph schema:\n\n%s\n", asr.newDgraphSchema)
+	}
+
+	err = asr.dgraph.Alter(ctx, &dgoapi.Operation{Schema: asr.newDgraphSchema})
+	if err != nil {
+		return nil, nil, schema.GQLWrapf(err,
+			"succeeded in saving GraphQL schema but failed to alter Dgraph schema "+
+				"(you should retry)")
+	}
+
+	resolverFactory := resolve.NewResolverFactory().
+		WithConventionResolvers(asr.newGQLSchema, asr.fns)
+
+	if asr.withIntrospection {
+		resolverFactory.WithSchemaIntrospection()
+	}
+
+	resolvers := resolve.New(asr.newGQLSchema, resolverFactory)
+	asr.gqlServer.ServeGQL(resolvers)
+
+	return assigned, mutated, nil
+}
+
+func getSchemaInput(m schema.Mutation) (string, error) {
+	input, ok := m.ArgValue(schema.InputArgName).(map[string]interface{})
+	if !ok {
+		return "", errors.New("couldn't get input argument")
+	}
+
+	return input["schema"].(string), nil
```

Comment:
I'm not 100% sure how to deal with args just yet.  They always come out of request parsing as `interface{}`, and kinda need to go back into the gql layer like that: e.g. I had to do

`m.SetArgTo(schema.InputArgName, map[string]interface{}{"schema": sch, "date": time.Now()})`

above to return the changed values back into the mutation. 

Different golang graphql libs have different ways of dealing with this, ranging from making you write casts to get what you need, to generating code that does the conversions to structs for you.

I wonder if we can have `type schemaInput struct { ... }` and get the `interface{}` to that type via json marshal->unmarshal and then back again in `SetArgTo`.  That's allow us type safe handling here in the couple of cases where we need it, without having to have much code around.